### PR TITLE
Graph scatter labels

### PR DESCRIPTION
**Pull request by normaljosh**
Added point labels and option to highlight for graphscatter and graphgroupscatter. 

For graphgroupscatter, we may want to consider having a highlight_group param which does what the highlight param does now, (pass it a list of groups to highlight) . That way we could have the highlight function operate the same way it does with a non-grouped scatter (pass it list of individual label items to show labels on and highlight).

### CODE REVIEW

File: sci_analysis/graphs/vector.py [hunk 1]
```diff
@@ -262,19 +262,31 @@ def __init__(self, xdata, ydata=None, **kwargs):
...
         self._title = kwargs['title'] if 'title' in kwargs else 'Bivariate'
         self._save_to = kwargs['save_to'] if 'save_to' in kwargs else None
         yname = kwargs['yname'] if 'yname' in kwargs else 'y Data'
         xname = kwargs['xname'] if 'xname' in kwargs else 'x Data'
+        if 'labels' in kwargs:
+            if len(kwargs['labels']) != len(xdata):
+                raise AttributeError('The length of passed labels does not match the length of xdata and ydata')
+            if ndarray.any(kwargs['labels'].index != xdata.data.index):
+                raise AttributeError('The index of passed labels does not match the index of xdata and ydata')
+        if 'highlight' in kwargs:
```
Comment (cmmorrow) [inline sci_analysis/graphs/vector.py]: Both of these if statements can be combined into a single statement.

File: sci_analysis/graphs/vector.py [hunk 2]
```diff
@@ -262,19 +262,31 @@ def __init__(self, xdata, ydata=None, **kwargs):
...
         self._contour_props = (31, 1.1)
         self._boxplot_borders = kwargs['boxplot_borders'] if 'boxplot_borders' in kwargs else False
         self._title = kwargs['title'] if 'title' in kwargs else 'Bivariate'
         self._save_to = kwargs['save_to'] if 'save_to' in kwargs else None
         yname = kwargs['yname'] if 'yname' in kwargs else 'y Data'
         xname = kwargs['xname'] if 'xname' in kwargs else 'x Data'
+        if 'labels' in kwargs:
+            if len(kwargs['labels']) != len(xdata):
+                raise AttributeError('The length of passed labels does not match the length of xdata and ydata')
+            if ndarray.any(kwargs['labels'].index != xdata.data.index):
```
Comment (cmmorrow) [inline sci_analysis/graphs/vector.py]: If drop_nan(), which runs as part of the Vector object constructor, removes values from the internal pandas DataFrame, the index and length might not match kwargs['labels']. One thought is to join labels on index to the DataFrame which will remove any index values that are not in the DataFrame. You would definitely need to write a  unit test to make sure it works correctly.

[the remark this answers is not in the document]
Comment (normaljosh) [reply sci_analysis/graphs/vector.py]: Checking the index comes after checking if the label argument is passed, so it shouldn't be calling .index if the label argument isn't passed. Either way, my latest update is attempting to remove the need to check the index at all.

File: sci_analysis/graphs/vector.py (hunk 1 above)
Comment (normaljosh) [reply sci_analysis/graphs/vector.py]: Combined these in the latest update

File: sci_analysis/graphs/vector.py (hunk 2 above)
Comment (normaljosh) [reply sci_analysis/graphs/vector.py]: I'm trying to work around this by adding a "dropped_vals" attribute to the numeric constructor, so when it runs drop_nan, it also creates a mask of what was dropped. Then, we can remove those values from the labels column, whether the labels are a pandas series, or a numpy array. 

It seems to work fine when I test it locally, but unit tests don't seem to recall the "dropped_vals"attribute  when it comes to run graphgroupscatter, they say the labels array doesn't have that attribute.

File: sci_analysis/graphs/vector.py [hunk 3]
```diff
@@ -262,19 +263,32 @@ def __init__(self, xdata, ydata=None, **kwargs):
...
         self._contours = kwargs['contours'] if 'contours' in kwargs else False
         self._contour_props = (31, 1.1)
         self._boxplot_borders = kwargs['boxplot_borders'] if 'boxplot_borders' in kwargs else False
         self._title = kwargs['title'] if 'title' in kwargs else 'Bivariate'
         self._save_to = kwargs['save_to'] if 'save_to' in kwargs else None
         yname = kwargs['yname'] if 'yname' in kwargs else 'y Data'
         xname = kwargs['xname'] if 'xname' in kwargs else 'x Data'
+        if 'labels' in kwargs:
+            #filter out values if drop_nan was used
+            self._labels = self._labels[~xdata._dropped_vals]
```
Comment (cmmorrow) [inline sci_analysis/graphs/vector.py]: xdata might or might not be a Vector object at this point in the constructor. If it isn't a Vector object, you'll get an AttributeError which is why your unit tests are failing. I think you can fix this by moving these label and highlight setup lines to after the call to super(), or by moving them into draw().

Good job with the unit tests by the way. The silver lining with them failing at the moment is they caught a real use case that should be addressed, and will result in more robust code.

File: sci_analysis/graphs/vector.py [hunk 4]
```diff
@@ -432,21 +467,32 @@ def __init__(self, xdata, ydata=None, groups=None, **kwargs):
...
+        self._labels = kwargs['labels'] if 'labels' in kwargs else [False]
         self._highlight = kwargs['highlight'] if 'highlight' in kwargs else None
         self._boxplot_borders = kwargs['boxplot_borders'] if 'boxplot_borders' in kwargs else True
         self._title = kwargs['title'] if 'title' in kwargs else 'Group Bivariate'
         self._save_to = kwargs['save_to'] if 'save_to' in kwargs else None
         yname = kwargs['yname'] if 'yname' in kwargs else 'y Data'
         xname = kwargs['xname'] if 'xname' in kwargs else 'x Data'
+        if 'labels' in kwargs:
+            #filter out values if drop_nan was used
+            print(xdata)
```
Comment (cmmorrow) [inline sci_analysis/graphs/vector.py]: Don't forget to remove this print statement.

File: sci_analysis/test/test_graph_groupscatter.py [hunk 5]
```diff
@@ -372,6 +372,40 @@ def test_27_scatter_two_groups_negative_corr(self):
...
+        input_1_x = st.norm.rvs(size=100)
+        input_1_y = [x + st.norm.rvs(0, 0.5, size=1)[0] for x in input_1_x]
+        input_2_x = st.norm.rvs(size=100)
+        input_2_y = [(x / 2) + st.norm.rvs(0, 0.2, size=1)[0] for x in input_2_x]
+        input_labels_array = np.random.choice(list('ABCDE'),  size=(200))
+        grp = [1] * 100 + [2] * 100
+        cs_x = np.concatenate((input_1_x, input_2_x))
+        cs_y = np.concatenate((input_1_y, input_2_y))
+        input_array = pd.DataFrame({'a': cs_x, 'b': cs_y, 'c': grp})
+        self.assertTrue(analyze(input_array['a'], input_array['b'], groups=input_array['c'], labels=input_labels_array,
```
Comment (cmmorrow) [inline sci_analysis/test/test_graph_groupscatter.py]: test_graph_groupscatter.py doesn't import analyze(). If you swap it out for GraphGroupScatter() I think the test will pass.

File: sci_analysis/data/numeric.py [hunk 6]
```diff
@@ -142,6 +142,7 @@ def drop_nan(self):
         arr : pandas.DataFrame
             A copy of the Numeric object's internal Series with all NaN values removed.
         """
+        self._dropped_vals = self._values[self._dep].isnull()
```
Comment (cmmorrow) [inline sci_analysis/data/numeric.py]: Should this be `self._values[self._ind].isnull()`? In drop_nan(), the dependent column is probably nan for each value.

File: sci_analysis/test/test_graph_groupscatter.py [hunk 7]
```diff
@@ -372,6 +372,48 @@ def test_27_scatter_two_groups_negative_corr(self):
...
+                                          save_to='{}test_group_scatter_29'.format(self.save_path)))
+
+
+    def test_30_groupscatter_dataframe(self):
+        """tests graphscater with dataframe input"""
+        np.random.seed(987654321)
+        df = pd.DataFrame(np.random.randn(100, 2), columns=list('xy'))
+        df['labels'] = np.random.choice(list('ABCDE'), len(df)).tolist()
+        df['groups'] = np.random.choice(list('XYZ'), len(df)).tolist()
+        self.assertTrue(GraphGroupScatter(df['x'], df['y'], groups=df['groups'], labels= df['labels'], save_to='{}test_scatter_135'.format(self.save_path)))
```
Comment (cmmorrow) [inline sci_analysis/test/test_graph_groupscatter.py]: The filename in save_to doesn't match the naming convention in the other tests (test_group_scatter_30).

File: sci_analysis/test/test_graph_scatter.py [hunk 8]
```diff
@@ -321,6 +321,40 @@ def test_132_negative_corr(self):
...
+        input_y_array = st.norm.rvs(size=2000)
+        input_labels_array = np.random.choice(list('ABCDE'),  size=(2000))
+        self.assertTrue(GraphScatter(input_x_array, input_y_array, labels= input_labels_array, highlight=['E'], save_to='{}test_scatter_135'.format(self.save_path)))
+
+    def test_136_Graphscatter_dataframe(self):
+        """tests graphscater with dataframe input"""
+        np.random.seed(987654321)
+        df = pd.DataFrame(np.random.randn(100, 2), columns=list('xy'))
+        df['labels'] = np.random.choice(list('ABCDE'), len(df)).tolist()
+        self.assertTrue(GraphScatter(df['x'], df['y'], labels= df['labels'], save_to='{}test_scatter_135'.format(self.save_path)))
```
Comment (cmmorrow) [inline sci_analysis/test/test_graph_scatter.py]: This filename also needs to be fixed. It should be 136 instead of 135.

**Comment by cmmorrow** [comment]
Looks good. I'll merge it.

**Pull request merged**